includes/class-wp-security-bp-files.php

```diff
@@ -0,0 +1,266 @@
+ * @author     Cyan Lovers <hello@cyanlove.com>
+ */
+class WP_Security_BP_Files {
+
+	/**
+	 * The unique identifier of this plugin.
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $plugin_name    The string used to uniquely identify this plugin.
+	 */
+	protected $plugin_name;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $request_uri    The URI where the request is made from
+	 */
+	protected $request_uri;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $nonce_action_name    The action for the nonce creation
+	 */
+	protected $nonce_action_name;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $wp_config    The name of the file wp-config.php
+	 */
+	protected $wp_config;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $htaccess    The name of the file .htaccess
+	 */
+	protected $htaccess;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      object    $wp_filesystem    The Object $wp_filesystem
+	 */
+	protected $wp_filesystem;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $root    The uri of the installation root
+	 */
+	protected $root;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      string    $parent_root    The uri of the parent installation root
+	 */
+	protected $parent_root;
+
+	/**
+	 * Short desc
+	 *
+	 * @since    1.0.0
+	 * @access   protected
+	 * @var      array    $json    The array that will be passed as a JSON file to the admin
+	 */
+	protected $json;
+
+	/**
+	 * Initialize the class and set its properties.
+	 *
+	 * @since    1.0.0
+	 * @param    string    $plugin_name       The name of this plugin.
+	 * @param    string    $request_uri       The URI from the request is made.
+	 */
+	public function __construct( $plugin_name, $request_uri ) {
+
+		$this->plugin_name = $plugin_name;
+		$this->request_uri = $request_uri;
+		$this->nonce_action_name = 'wp-security-bp-file-access';
+		$this->wp_config = 'wp-config.php';
+		$this->json = array(
+			'status'   => 'fail',
+			'message'  => '',
+			'button'   => false,
+			'uri'      => '',
+		);
+
+		$access_type = get_filesystem_method();
+		if ( $access_type === 'direct' ) {
+
+			// request credentials
+			$creds = $this->request_credentials();
+
+			// if user has no permission is asked for ftp credentials
+			if ( false === $creds ) {
+				return; // stop processing here
+			}
+
+			// if ftp credentials are not ok user is asked again for ftp credentials
+			if ( false === WP_Filesystem( $creds ) ) {
+				$this->request_credentials( true );
+				return; // stop processing here
+			}
+
+			// call global $wp_filesystem variable
+			global $wp_filesystem;
+			$this->wp_filesystem = $wp_filesystem;
```

Comment:
why do we need this reassign?

Reply:
Because `global` just makes the variable available in the current scope (in this case `__construct()`) and I need to access to this variable from more methods in this class. I could call it with `global` on every method I need but I found more useful to have it accessible through a propriety.